RHDEVDOCS-4781 - Loki (#59285)
openshift/openshift-docs: modules/logging-loki-storage.adoc

== AWS S3

.Prerequisites
* Loki Operator deployed
* Create a link:https://docs.aws.amazon.com/AmazonS3/latest/userguide/create-bucket-overview.html[bucket] on AWS.

.Procedure
* Create an Object Storage secret with keys as follows:

[source,terminal]
----
oc create secret generic logging-loki-s3 \
  --from-literal=bucketnames="<BUCKET_NAME>" \
  --from-literal=endpoint="<AWS_BUCKET_ENDPOINT>" \
  --from-literal=access_key_id="<AWS_ACCESS_KEY_ID>" \
  --from-literal=access_key_secret="<AWS_ACCESS_KEY_SECRET>" \
  --from-literal=region="<AWS_REGION_OF_YOUR_BUCKET>"
----

** Suggested secret name: `logging-loki-s3`

** Secret type: `s3`

== Azure

.Prerequisites
* Loki Operator deployed
* Create a link:https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction[bucket] on Azure.

.Procedure
* Create an Object Storage secret with keys as follows:

[source,terminal]
----
oc create secret generic logging-loki-azure \
  --from-literal=container="<AZURE_CONTAINER_NAME>" \
  --from-literal=environment="<AZURE_ENVIRONMENT>" \
  --from-literal=account_name="<AZURE_ACCOUNT_NAME>" \
  --from-literal=account_key="<AZURE_ACCOUNT_KEY>"

> ----

Reviewer: Can you provide the example for AZURE_ENVIRONMENT as below?
# Azure Cloud environment. Supported values are: AzureGlobal, AzureChinaCloud,
# AzureGermanCloud, AzureUSGovernment.

. Create an Object Storage secret with keys `bucketname` and `key.json` as follows:

+
[source,terminal]
----
oc create secret generic logging-loki-gcs \
  --from-literal=bucketname="<BUCKET_NAME>" \
  --from-file=key.json="<PATH/TO/KEY.JSON>"
----

Where `logging-loki-gcs` is the secret name, `<BUCKET_NAME>` is the name of bucket created in prerequisites step and `<PATH/TO/KEY.JSON>` is the file path where the `key.json` was copied to.

** Suggested secret name: `logging-loki-gcs`

** Secret type: `gcs`

== Minio

.Prerequisites

* Loki Operator deployed

* link:https://operator.min.io/[Minio] deployed on your Cluster.

* Create a link:https://docs.min.io/docs/minio-client-complete-guide.html[bucket] on Minio via CLI.

.Procedure

* Create an Object Storage secret with keys as follows:

[source,terminal]
----
oc create secret generic logging-loki-minio \
  --from-literal=bucketnames="<BUCKET_NAME>" \
  --from-literal=endpoint="<MINIO_BUCKET_ENDPOINT>" \
  --from-literal=access_key_id="<MINIO_ACCESS_KEY_ID>" \
  --from-literal=access_key_secret="<MINIO_ACCESS_KEY_SECRET>"
----

** Suggested secret name: `logging-loki-minio`

** Secret type: `s3`

> == OpenShift Data Foundation

Author: @anpingli

.Prerequisites

* Loki Operator deployed

* Deploy the [OpenShift Data Foundation](https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.12) on your cluster.

.Procedure

. Create an ObjectBucketClaim in `openshift-logging` namespace:
+
[source,yaml]
----
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucketClaim
metadata:
  name: loki-bucket-odf
  namespace: openshift-logging
spec:

>   generateBucketName: loki-bucket-odf

Reviewer 2: When I try to create ObjectBucketClaim with above yaml, I get below error:
```
$ cat << EOF | oc create -f -
> apiVersion: objectbucket.io/v1alpha1
> kind: ObjectBucketClaim
> metadata:
>   name: loki-bucket-odf
>   namespace: openshift-logging
> spec:
>   generateBucketName: loki-bucket-odf
> EOF
The ObjectBucketClaim "loki-bucket-odf" is invalid: spec.storageClassName: Required value
```

Author: @periklis -- Can you assist with the above please?

Author: @periklis ?

Reviewer 3: I can confirm that `storageClassName` is a required field for this resource. I have [found an example](https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.13/html/managing_hybrid_and_multicloud_resources/object-bucket-claim#dynamic-object-bucket-claim_rhodf) that uses `openshift-storage.noobaa.io` which looks like it might be a "constant value" for ODF. I have not run a test cluster with ODF on it in a while, so I unfortunately don't remember anymore what value I used for my `ObjectBucketClaim`.

----

. Get bucket properties from the associated ConfigMap:
+
[source,terminal]
----
BUCKET_HOST=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_HOST}')
BUCKET_NAME=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_NAME}')
BUCKET_PORT=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_PORT}')
----

. Get bucket access key from the associated Secret:
+
[source,terminal]
----
ACCESS_KEY_ID=$(oc get -n openshift-logging secret loki-bucket-odf -o jsonpath='{.data.AWS_ACCESS_KEY_ID}' | base64 -d)
SECRET_ACCESS_KEY=$(oc get -n openshift-logging secret loki-bucket-odf -o jsonpath='{.data.AWS_SECRET_ACCESS_KEY}' | base64 -d)
----

.. Create an Object Storage secret with keys as follows:
+
[source,terminal]
----
oc create -n openshift-logging secret generic logging-loki-odf \
--from-literal=access_key_id="${ACCESS_KEY_ID}" \
--from-literal=access_key_secret="${SECRET_ACCESS_KEY}" \
--from-literal=bucketnames="${BUCKET_NAME}" \

> --from-literal=endpoint="https://${BUCKET_HOST}:${BUCKET_PORT}"

Reviewer 2: One more thing we should mention here is: if the schema of endpoint is `https`, users should:
1. create a configmap with below yaml in the namespace where the lokistack will be in(here it's `openshift-logging`):
```
cat << EOF | oc create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: lokistack-odf-ca
  namespace: openshift-logging
  annotations:
    service.beta.openshift.io/inject-cabundle: "true"
EOF
```
note: the configmap name can be changed.
2. users should add below lines into the lokistack CR:
```
spec:
  storage:
    secret:
      name: logging-loki-odf
      type: s3
    tls:
      caName: lokistack-odf-ca
```
Or the loki pods may not be ready and raise some `x509` errors.

If the schema is `http`, then no need to do above steps.

// Module is included in the following assemblies:
// logging/logging-loki-ocp.adoc
//
:_content-type: PROCEDURE
[id="logging-loki-storage_{context}"]
= Loki log storage

> The Loki Operator supports link:https://aws.amazon.com/[AWS S3], as well as other S3 compatible object stores such as link:https://min.io/[Minio] and link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[OpenShift Data Foundation]. link:https://azure.microsoft.com[Azure], link:https://cloud.google.com/[GCS], and link:https://docs.openstack.org/swift/latest/[Swift] are also supported.

Reviewer 4: ```suggestion
The Loki Operator supports link:https://aws.amazon.com/[AWS S3], as well as other S3 compatible object stores such as link:https://min.io/[Minio] and link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[OpenShift Data Foundation], link:https://azure.microsoft.com[Azure], link:https://cloud.google.com/[GCS], and link:https://docs.openstack.org/swift/latest/[Swift] are also supported.
```

// Module is included in the following assemblies:
// logging/logging-loki-ocp.adoc
//
:_content-type: PROCEDURE
[id="logging-loki-storage_{context}"]
= Loki log storage

> The Loki Operator supports link:https://aws.amazon.com/[AWS S3], link:https://azure.microsoft.com[Azure], link:https://cloud.google.com/[GCS], link:https://min.io/[Minio],  link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[OpenShift Data Foundation] and link:https://docs.openstack.org/swift/latest/[Swift] for LokiStack object storage.

Reviewer 5: Do we need to add Alibaba cloud here too?

> The Loki Operator supports link:https://aws.amazon.com/[AWS S3], link:https://azure.microsoft.com[Azure], link:https://cloud.google.com/[GCS], link:https://min.io/[Minio],  link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[OpenShift Data Foundation], link:https://us.alibabacloud.com/[Alibaba Cloud] and link:https://docs.openstack.org/swift/latest/[Swift] for LokiStack object storage.

Reviewer 3: Does it make sense to group the "S3-compatible" storages (AWS, Minio, ODF, Alibaba-Cloud) together, so that it becomes visible that we're not explicitly supporting _these storages_ but can work with anything that has a S3-compatible API (in addition to some others like GCS which _do_ have a different API).

The recommended nomenclature for Loki storage is `logging-loki-<type>`. General secret creation format is as follows:
//Generic secret creation
include::snippets/logging-create-secret-snip.adoc[]

You can verify a secret has been created by running the following command:

[source,terminal]
----

> oc get secrets

Reviewer 3: Does this need to mention our namespace (`-n openshift-logging`) or can we assume that the user knows where they created the secret?

----

The table below displays the `type` values within the `LokiStack` CR for each storage provider. For more detailed information, consult the section for your storage provider.

[options="header"]
.Secret type quick reference
|================================================
| Storage provider          | Secret `type` value
| AWS                       | s3
| Azure                     | azure
| Google Cloud              | gcs
| Minio                     | s3
| OpenShift Data Foundation | s3
| Swift                     | swift

> | Alibaba Cloud             | alibabacloud

Reviewer: loki-operator doesn't support type:alibabacloud,  can we remove alicould from the doc?  https://github.com/openshift/loki/blob/main/operator/bundle/openshift/manifests/loki.grafana.com_lokistacks.yaml#L4285

Reviewer: @periklis  Do we plan to add alibabacloud in future(Logging 5.8)?

Reviewer 4: It is already into the code base. I just do not recall if we have a QE approval on this. So let's skip this for now.

Author: Already removed. :)

----

Optionally you can provide project-specific data and/or a region as follows:

[source,terminal]
----
oc create secret generic logging-loki-swift \
  --from-literal=auth_url="<SWIFT_AUTH_URL>" \
  --from-literal=username="<SWIFT_USERNAMEClaim>" \
  --from-literal=user_domain_name="<SWIFT_USER_DOMAIN_NAME>" \
  --from-literal=user_domain_id="<SWIFT_USER_DOMAIN_ID>" \
  --from-literal=user_id="<SWIFT_USER_ID>" \
  --from-literal=password="<SWIFT_PASSWORD>" \
  --from-literal=domain_id="<SWIFT_DOMAIN_ID>" \
  --from-literal=domain_name="<SWIFT_DOMAIN_NAME>" \
  --from-literal=container_name="<SWIFT_CONTAINER_NAME>" \
  --from-literal=project_id="<SWIFT_PROJECT_ID>" \
  --from-literal=project_name="<SWIFT_PROJECT_NAME>" \
  --from-literal=project_domain_id="<SWIFT_PROJECT_DOMAIN_ID>" \
  --from-literal=project_domain_name="<SWIFT_PROJECT_DOMAIN_name>" \
  --from-literal=region="<SWIFT_REGION>" \
----

** Suggested secret name: `logging-loki-swift`

** Secret type: `swift`

== Alibaba Cloud

.Prerequisites
* Loki Operator deployed
* Create a link:https://www.alibabacloud.com/help/en/object-storage-service/latest/create-buckets-2[bucket] on AlibabaCloud.

.Procedure

> * Create an Object Storage secret with keys as follows:

Reviewer: Until Logging 5.7.2, alicloud isn't supported.

// Module is included in the following assemblies:
// logging/logging-loki-ocp.adoc
//
:_content-type: PROCEDURE
[id="logging-loki-storage_{context}"]

= Loki log storage

The Loki Operator supports link:https://aws.amazon.com/[AWS S3], link:https://azure.microsoft.com[Azure], link:https://cloud.google.com/[GCS], link:https://min.io/[Minio],  link:https://www.redhat.com/en/technologies/cloud-computing/openshift-data-foundation[OpenShift Data Foundation] and link:https://docs.openstack.org/swift/latest/[Swift] for LokiStack object storage.

> The recommended nomenclature for Loki storage is `logging-loki-<type>`. General secret creation format is as follows:

Reviewer 6: ```suggestion
The recommended nomenclature for Loki storage is `logging-loki-<type>`. The general format for creating secrets is as follows:
```

  --from-literal=region="<AWS_REGION_OF_YOUR_BUCKET>"
----

Suggested secret name: `logging-loki-s3`

type: `s3`

== Azure

.Prerequisites
* Loki Operator deployed
* Create a link:https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction[bucket] on Azure.

.Procedure
* Create an Object Storage secret with keys as follows:

[source,terminal]
----
oc create secret generic logging-loki-azure \
  --from-literal=container="<AZURE_CONTAINER_NAME>" \
  --from-literal=environment="<AZURE_ENVIRONMENT>" \
  --from-literal=account_name="<AZURE_ACCOUNT_NAME>" \
  --from-literal=account_key="<AZURE_ACCOUNT_KEY>"
----

Suggested secret name: `logging-loki-azure`

type: `azure`

== Google Cloud Storage

.Prerequisites

* Loki Operator deployed

* Create a link:https://cloud.google.com/resource-manager/docs/creating-managing-projects[project] on Google Cloud Platform.

* Create a link:https://cloud.google.com/storage/docs/creating-buckets[bucket] under same project.

* Create a link:https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account[service account] under same project for GCP authentication.

.Procedure

> 

Reviewer 6: Please format as a multi-step procedure with line numbers (`. `) and `+` characters to align following text.

* Copy the service account credentials received from GCP into a file name `key.json`.

* Create an Object Storage secret with keys `bucketname` and `key.json` as follows:

[source,terminal]
----
oc create secret generic logging-loki-gcs \
  --from-literal=bucketname="<BUCKET_NAME>" \
  --from-file=key.json="<PATH/TO/KEY.JSON>"
----

> Where `logging-loki-gcs` is the secret name, `<BUCKET_NAME>` is the name of bucket created in prerequisites step and `<PATH/TO/KEY.JSON>` is the file path where the `key.json` was copied to.

Reviewer 6: ```suggestion
Where `logging-loki-gcs` is the secret name, `<BUCKET_NAME>` is the name of bucket created in the prerequisites step, and `<PATH/TO/KEY.JSON>` is the file path where the `key.json` was copied to.
```

Suggested secret name: `logging-loki-gcs`

type: `gcs`

== Minio

.Prerequisites

* Loki Operator deployed

* link:https://operator.min.io/[Minio] deployed on your Cluster.

* Create a link:https://docs.min.io/docs/minio-client-complete-guide.html[bucket] on Minio via CLI.

.Procedure

* Create an Object Storage secret with keys as follows:

> 

Reviewer 6: ```suggestion
+
```

[source,terminal]
----
oc create secret generic logging-loki-minio \
  --from-literal=bucketnames="<BUCKET_NAME>" \
  --from-literal=endpoint="<MINIO_BUCKET_ENDPOINT>" \
  --from-literal=access_key_id="<MINIO_ACCESS_KEY_ID>" \
  --from-literal=access_key_secret="<MINIO_ACCESS_KEY_SECRET>"
----

Suggested secret name: `logging-loki-minio`

> type: `s3`

Reviewer 6: Maybe rephrase these as sentences.

== OpenShift Data Foundation

.Prerequisites

* Loki Operator deployed

* Deploy the [OpenShift Data Foundation](https://access.redhat.com/documentation/en-us/red_hat_openshift_data_foundation/4.12) on your cluster.

.Procedure

* Create an ObjectBucketClaim in `openshift-logging` namespace:

[source,yaml]
----
apiVersion: objectbucket.io/v1alpha1
kind: ObjectBucketClaim
metadata:
  name: loki-bucket-odf
  namespace: openshift-logging
spec:
  generateBucketName: loki-bucket-odf
----

* Get bucket properties from the associated ConfigMap:

[source,terminal]
----
BUCKET_HOST=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_HOST}')
BUCKET_NAME=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_NAME}')
BUCKET_PORT=$(oc get -n openshift-logging configmap loki-bucket-odf -o jsonpath='{.data.BUCKET_PORT}')
----

> * Get bucket access key from the associated Secret:

Reviewer 6: ```suggestion
* Get bucket access key from the associated secret:
```
to match the other instances of `secret`.

[source,terminal]
----
ACCESS_KEY_ID=$(oc get -n openshift-logging secret loki-bucket-odf -o jsonpath='{.data.AWS_ACCESS_KEY_ID}' | base64 -d)
SECRET_ACCESS_KEY=$(oc get -n openshift-logging secret loki-bucket-odf -o jsonpath='{.data.AWS_SECRET_ACCESS_KEY}' | base64 -d)
----

** Create an Object Storage secret with keys as follows:
[source,terminal]
----
oc create -n openshift-logging secret generic logging-loki-odf \
--from-literal=access_key_id="${ACCESS_KEY_ID}" \
--from-literal=access_key_secret="${SECRET_ACCESS_KEY}" \
--from-literal=bucketnames="${BUCKET_NAME}" \
--from-literal=endpoint="https://${BUCKET_HOST}:${BUCKET_PORT}"
----

Where `logging-loki-odf` is the secret name. The values for `ACCESS_KEY_ID`, `SECRET_ACCESS_KEY`, `BUCKET_NAME`, `BUCKET_HOST` and `BUCKET_PORT` are taken from your ObjectBucketClaim's accompanied secret and ConfigMap.

Suggested secret name: `logging-loki-odf`

type: `s3`

== Swift

.Prerequisites

* Loki Operator deployed

* Create a https://docs.openstack.org/newton/user-guide/cli-swift-create-containers.html[bucket] on Swift.

.Procedure

* Create an Object Storage secret with keys as follows:

> 

Reviewer 6: ```suggestion
+
```

[source,terminal]
----
oc create secret generic logging-loki-swift \
  --from-literal=auth_url="<SWIFT_AUTH_URL>" \
  --from-literal=username="<SWIFT_USERNAMEClaim>" \
  --from-literal=user_domain_name="<SWIFT_USER_DOMAIN_NAME>" \
  --from-literal=user_domain_id="<SWIFT_USER_DOMAIN_ID>" \
  --from-literal=user_id="<SWIFT_USER_ID>" \
  --from-literal=password="<SWIFT_PASSWORD>" \
  --from-literal=domain_id="<SWIFT_DOMAIN_ID>" \
  --from-literal=domain_name="<SWIFT_DOMAIN_NAME>" \
  --from-literal=container_name="<SWIFT_CONTAINER_NAME>" \
----

> 

Reviewer 6: ```suggestion
+
```

Where `logging-loki-swift` is the secret name.

> Optionally you can provide project specific data and/or a region as follows:

Reviewer 6: ```suggestion
Optionally you can provide project-specific data and/or a region as follows:
```

> Optionally you can provide project specific data and/or a region as follows:

Reviewer 6: ```suggestion
Optionally, you can provide project-specific data and/or a region as follows:
```

[source,terminal]
----
oc create secret generic logging-loki-swift \
  --from-literal=auth_url="<SWIFT_AUTH_URL>" \
  --from-literal=username="<SWIFT_USERNAMEClaim>" \
  --from-literal=user_domain_name="<SWIFT_USER_DOMAIN_NAME>" \
  --from-literal=user_domain_id="<SWIFT_USER_DOMAIN_ID>" \
  --from-literal=user_id="<SWIFT_USER_ID>" \
  --from-literal=password="<SWIFT_PASSWORD>" \
  --from-literal=domain_id="<SWIFT_DOMAIN_ID>" \
  --from-literal=domain_name="<SWIFT_DOMAIN_NAME>" \
  --from-literal=container_name="<SWIFT_CONTAINER_NAME>" \
  --from-literal=project_id="<SWIFT_PROJECT_ID>" \
  --from-literal=project_name="<SWIFT_PROJECT_NAME>" \
  --from-literal=project_domain_id="<SWIFT_PROJECT_DOMAIN_ID>" \
  --from-literal=project_domain_name="<SWIFT_PROJECT_DOMAIN_name>" \
  --from-literal=region="<SWIFT_REGION>" \
----

Suggested secret name: `logging-loki-swift`

type: `swift`

> == AlibabaCloud

Reviewer 6: ```suggestion
== Alibaba Cloud
```

.Prerequisites
* Loki Operator deployed

> * Create a link:https://www.alibabacloud.com/help/en/object-storage-service/latest/create-buckets-2[bucket] on AlibabaCloud.

Reviewer 6: ```suggestion
* Create a link:https://www.alibabacloud.com/help/en/object-storage-service/latest/create-buckets-2[bucket] on Alibaba Cloud.
```

.Procedure

> * Deploy the Loki Operator to your cluster.
> 
> * Create an Object Storage secret with keys as follows:
> 
> [source,terminal]

Reviewer 6: ```suggestion
. Deploy the Loki Operator to your cluster.

. Create an Object Storage secret with keys as follows:
+
[source,terminal]
```
